File: internal/integrations/bitbucket/actions.go
package bitbucket

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/roee-hersh/hallpass/internal/catalog"
	"github.com/roee-hersh/hallpass/internal/integration"
)

// level is an effective permission on a repository or project, ordered.
type level int

const (
	levelNone level = iota
	levelRead
	levelWrite
	levelAdmin
)

func (l level) String() string {
	switch l {
	case levelRead:
		return "read"
	case levelWrite:
		return "write"
	case levelAdmin:
		return "admin"
	}
	return "none"
}

// action is one named question.
type action struct {
	name, desc string
	// resource is the type the action takes: repo, project or workspace.
	resource string
	// level is the repository or project permission needed.
	level level
	// branch is the branch-restriction kind consulted when the repo
	// resource names a branch: "push" or "merge".
	branch string
	// createRepo marks project-level repository creation, which Cloud
	// grants separately from write.
	createRepo bool
	// role is the workspace role needed: "member" or "admin".
	role string
}

var actionList = []action{
	{name: "repo.read", desc: "read and clone the repository; needs read", resource: "repo", level: levelRead},
	{name: "repo.push", desc: "push to the repository (or to @branch, checked against its branch restrictions); needs write", resource: "repo", level: levelWrite, branch: "push"},
	{name: "pr.merge", desc: "merge a pull request (into @branch, checked against its branch restrictions); needs write", resource: "repo", level: levelWrite, branch: "merge"},
	{name: "repo.admin", desc: "administer the repository: settings, permissions, deletion; needs admin", resource: "repo", level: levelAdmin},
	{name: "project.read", desc: "see the project and read its repositories; needs read", resource: "project", level: levelRead},
	{name: "project.write", desc: "push to the project's repositories; needs write", resource: "project", level: levelWrite},
	{name: "repo.create", desc: "create a repository in the project; needs create-repo (Cloud) or project admin (Data Center, see the doc)", resource: "project", level: levelWrite, createRepo: true},
	{name: "project.admin", desc: "administer the project; needs admin", resource: "project", level: levelAdmin},
	{name: "workspace.member", desc: "is a member of the workspace (Cloud) or a licensed user (Data Center)", resource: "workspace", role: "member"},
	{name: "workspace.admin", desc: "is a workspace owner (Cloud) or a global administrator (Data Center)", resource: "workspace", role: "admin"},
}

var actions = func() map[string]action {
	m := map[string]action{}
	for _, a := range actionList {
		m[a.name] = a
	}
	return m
}()

// Actions of the bitbucket integration.
func (Integration) Actions() []catalog.Action {
	out := make([]catalog.Action, 0, len(actionList))
	for _, a := range actionList {
		out = append(out, catalog.Action{Name: a.name, Description: a.desc + " (" + a.resource + ")"})
	}
	return out
}

var (
	// slugRe is a workspace slug, repository slug or project key. Cloud
	// slugs are lowercase with dots, hyphens and underscores; Data Center
	// project keys are upper case and personal projects start with "~".
	slugRe = regexp.MustCompile(`^~?[A-Za-z0-9][A-Za-z0-9._-]{0,127}$`)
	// uuidRe is a Cloud UUID in braces, also accepted where a slug is.
	uuidRe   = regexp.MustCompile(`^\{[0-9a-fA-F-]{36}\}$`)
	branchRe = regexp.MustCompile(`^[^\x00-\x20\x7f~^:?*\[\\]{1,255}$`)
)

func validSlug(s string) bool { return slugRe.MatchString(s) || uuidRe.MatchString(s) }

func validBranch(s string) bool {
	return branchRe.MatchString(s) && !strings.HasPrefix(s, "-") && !strings.HasPrefix(s, "/") &&
		!strings.HasSuffix(s, "/") && !strings.HasSuffix(s, ".") && !strings.HasSuffix(s, ".lock") &&
		!strings.Contains(s, "..") && !strings.Contains(s, "//") && !strings.Contains(s, "@{")
}

// target is a parsed resource.
type target struct {
	action action
	// project is the project key (Data Center, and Cloud projects).
	project string
	// repo is the repository slug; empty for project and workspace targets.
	repo string
	// branch is the branch named with @, if any.
	branch string
}

func invalid(format string, args ...any) error {
	return integration.Errorf(integration.CodeInvalidRequest, format, args...)
}

// parseTarget validates the resource for the action.
//
//	repo:<slug>[@branch]                Cloud, the workspace is the connection's
//	repo:<PROJECT>/<slug>[@branch]      Data Center
//	project:<key>
//	workspace                           the connection's workspace or instance
func parseTarget(actionName string, r catalog.Resource, dataCenter bool) (target, error) {
	a, ok := actions[actionName]
	if !ok {
		return target{}, invalid("unknown action %q", actionName)
	}
	if len(r.Query) > 0 {
		return target{}, invalid("resource %q must not carry a query", r.Raw)
	}
	if r.Type != a.resource {
		return target{}, invalid("action %s takes a %s: resource, not %s:", a.name, a.resource, r.Type)
	}
	t := target{action: a}
	switch r.Type {
	case "workspace":
		if r.ID != "" {
			return target{}, invalid("workspace takes no id: the connection names the workspace")
		}
	case "project":
		if !validSlug(r.ID) {
			return target{}, invalid("project: id must be a project key")
		}
		t.project = r.ID
	case "repo":
		id, branch := catalog.SplitBranch(r.ID)
		if strings.Contains(r.ID, "@") && branch == "" {
			return target{}, invalid("repo: a branch after @ must not be empty")
		}
		if branch != "" && !validBranch(branch) {
			return target{}, invalid("branch %q is not a valid branch name", branch)
		}
		if branch != "" && a.branch == "" {
			return target{}, invalid("action %s does not take a branch", a.name)
		}
		t.branch = branch
		if dataCenter {
			project, repo, ok := strings.Cut(id, "/")
			if !ok || !validSlug(project) || !validSlug(repo) {
				return target{}, invalid("repo: id must be <PROJECT>/<slug>[@branch] on Data Center")
			}
			t.project, t.repo = project, repo
		} else {
			if strings.Contains(id, "/") || !validSlug(id) {
				return target{}, invalid("repo: id must be <slug>[@branch]; the workspace is the connection's")
			}
			t.repo = id
		}
	}
	return t, nil
}

// String names the target for decision texts.
func (t target) String() string {
	switch t.action.resource {
	case "workspace":
		return "the workspace"
	case "project":
		return "project " + t.project
	}
	s := "repository " + t.repo
	if t.project != "" {
		s = "repository " + t.project + "/" + t.repo
	}
	if t.branch != "" {
		s += "@" + t.branch
	}
	return s
}

// globMatch matches a Bitbucket branch pattern against a branch name. "*"
// and "**" match any run of characters, including "/"; "?" one character.
// Anything else is literal. Patterns with character classes or
// alternations are reported unsupported.
func globMatch(pattern, name string) (matched, supported bool) {
	if strings.ContainsAny(pattern, "[]{}") {
		return false, false
	}
	return glob(pattern, name), true
}

func glob(p, s string) bool {
	for len(p) > 0 {
		switch p[0] {
		case '*':
			p = strings.TrimLeft(p, "*")
			if p == "" {
				return true
			}
			for i := 0; i <= len(s); i++ {
				if glob(p, s[i:]) {
					return true
				}
			}
			return false
		case '?':
			if s == "" {
				return false
			}
			p, s = p[1:], s[1:]
		default:
			if s == "" || p[0] != s[0] {
				return false
			}
			p, s = p[1:], s[1:]
		}
	}
	return s == ""
}

// refMatch matches a pattern that may carry a refs/heads/ prefix against a
// branch name, as Bitbucket does.
func refMatch(pattern, branch string) (bool, bool) {
	pattern = strings.TrimPrefix(pattern, "refs/heads/")
	return globMatch(pattern, branch)
}

func describeLevel(l level, needed level) string {
	if l >= needed {
		return fmt.Sprintf("has %s (needs %s)", l, needed)
	}
	return fmt.Sprintf("has %s, needs %s", l, needed)
}

File: hallpass-py/src/hallpass/integrations/bitbucket/actions.py
"""The bitbucket action table, resource parsing and branch pattern matching
(Go: internal/integrations/bitbucket/actions.go)."""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import IntEnum

from hallpass.core.catalog import Action, Resource, go_bytes, split_branch
from hallpass.core.decision import Code, HallpassError, errorf
from hallpass.core.errors import go_quote


class Level(IntEnum):
    """An effective permission on a repository or project, ordered."""

    NONE = 0
    READ = 1
    WRITE = 2
    ADMIN = 3

    def __str__(self) -> str:
        if self == Level.READ:
            return "read"
        if self == Level.WRITE:
            return "write"
        if self == Level.ADMIN:
            return "admin"
        return "none"


@dataclass(frozen=True)
class BBAction:
    """One named question."""

    name: str
    desc: str
    # The type the action takes: repo, project or workspace.
    resource: str
    # The repository or project permission needed.
    level: Level = Level.NONE
    # The branch-restriction kind consulted when the repo resource names a
    # branch: "push" or "merge".
    branch: str = ""
    # Project-level repository creation, which Cloud grants separately from
    # write.
    create_repo: bool = False
    # The workspace role needed: "member" or "admin".
    role: str = ""


ACTION_LIST: tuple[BBAction, ...] = (
    BBAction("repo.read", "read and clone the repository; needs read", "repo", Level.READ),
    BBAction(
        "repo.push",
        "push to the repository (or to @branch, checked against its branch restrictions); needs write",
        "repo",
        Level.WRITE,
        branch="push",
    ),
    BBAction(
        "pr.merge",
        "merge a pull request (into @branch, checked against its branch restrictions); needs write",
        "repo",
        Level.WRITE,
        branch="merge",
    ),
    BBAction("repo.admin", "administer the repository: settings, permissions, deletion; needs admin", "repo", Level.ADMIN),
    BBAction("project.read", "see the project and read its repositories; needs read", "project", Level.READ),
    BBAction("project.write", "push to the project's repositories; needs write", "project", Level.WRITE),
    BBAction(
        "repo.create",
        "create a repository in the project; needs create-repo (Cloud) or project admin (Data Center, see the doc)",
        "project",
        Level.WRITE,
        create_repo=True,
    ),
    BBAction("project.admin", "administer the project; needs admin", "project", Level.ADMIN),
    BBAction("workspace.member", "is a member of the workspace (Cloud) or a licensed user (Data Center)", "workspace", role="member"),
    BBAction("workspace.admin", "is a workspace owner (Cloud) or a global administrator (Data Center)", "workspace", role="admin"),
)

ACTIONS: dict[str, BBAction] = {a.name: a for a in ACTION_LIST}


def catalog_actions() -> list[Action]:
    return [Action(a.name, a.desc + " (" + a.resource + ")") for a in ACTION_LIST]


# SLUG_RE is a workspace slug, repository slug or project key. Cloud slugs
# are lowercase with dots, hyphens and underscores; Data Center project keys
# are upper case and personal projects start with "~".
SLUG_RE = re.compile(r"~?[A-Za-z0-9][A-Za-z0-9._-]{0,127}")
# UUID_RE is a Cloud UUID in braces, also accepted where a slug is.
UUID_RE = re.compile(r"\{[0-9a-fA-F-]{36}\}")
BRANCH_RE = re.compile(r"[^\x00-\x20\x7f~^:?*\[\\]{1,255}")


def valid_slug(s: str) -> bool:
    return SLUG_RE.fullmatch(s) is not None or UUID_RE.fullmatch(s) is not None


def valid_branch(s: str) -> bool:
    return (
        BRANCH_RE.fullmatch(s) is not None
        and not s.startswith("-")
        and not s.startswith("/")
        and not s.endswith("/")
        and not s.endswith(".")
        and not s.endswith(".lock")
        and ".." not in s
        and "//" not in s
        and "@{" not in s
    )


@dataclass(frozen=True)
class Target:
    """A parsed resource."""

    action: BBAction
    # The project key (Data Center, and Cloud projects).
    project: str = ""
    # The repository slug; empty for project and workspace targets.
    repo: str = ""
    # The branch named with @, if any.
    branch: str = ""

    def __str__(self) -> str:
        """Names the target for decision texts."""
        if self.action.resource == "workspace":
            return "the workspace"
        if self.action.resource == "project":
            return "project " + self.project
        s = "repository " + self.repo
        if self.project != "":
            s = "repository " + self.project + "/" + self.repo
        if self.branch != "":
            s += "@" + self.branch
        return s


def invalid(text: str) -> HallpassError:
    return errorf(Code.INVALID_REQUEST, text)


def parse_target(action_name: str, r: Resource, data_center: bool) -> Target:
    """Validate the resource for the action; raises an invalid_request
    HallpassError.

        repo:<slug>[@branch]                Cloud, the workspace is the connection's
        repo:<PROJECT>/<slug>[@branch]      Data Center
        project:<key>
        workspace                           the connection's workspace or instance
    """
    a = ACTIONS.get(action_name)
    if a is None:
        raise invalid(f"unknown action {go_quote(action_name)}")
    if r.query:
        raise invalid(f"resource {go_quote(r.raw)} must not carry a query")
    if r.type != a.resource:
        raise invalid(f"action {a.name} takes a {a.resource}: resource, not {r.type}:")
    project = repo = branch = ""
    if r.type == "workspace":
        if r.id != "":
            raise invalid("workspace takes no id: the connection names the workspace")
    elif r.type == "project":
        if not valid_slug(r.id):
            raise invalid("project: id must be a project key")
        project = r.id
    elif r.type == "repo":
        rid, branch = split_branch(r.id)
        if "@" in r.id and branch == "":
            raise invalid("repo: a branch after @ must not be empty")
        if branch != "" and not valid_branch(branch):
            raise invalid(f"branch {go_quote(branch)} is not a valid branch name")
        if branch != "" and a.branch == "":
            raise invalid(f"action {a.name} does not take a branch")
        if data_center:
            p, sep, rp = rid.partition("/")
            if not sep or not valid_slug(p) or not valid_slug(rp):
                raise invalid("repo: id must be <PROJECT>/<slug>[@branch] on Data Center")
            project, repo = p, rp
        else:
            if "/" in rid or not valid_slug(rid):
                raise invalid("repo: id must be <slug>[@branch]; the workspace is the connection's")
            repo = rid
    return Target(a, project, repo, branch)


def glob(p: str, s: str, crossing: bool) -> bool:
    """An Ant-style pattern match: "*" and "?" stay within one path segment,
    "**" crosses segments. With crossing set, "*" and "?" match "/" too.
    Patterns with character classes or alternations are not handled.
    Byte-wise, as Go indexes strings."""
    return _glob(go_bytes(p), go_bytes(s), crossing)


def _glob(p: bytes, s: bytes, crossing: bool) -> bool:
    slash = 0x2F
    while p:
        if p.startswith(b"**"):
            p = p.lstrip(b"*")
            if p.startswith(b"/"):
                p = p[1:]
            if not p:
                return True
            return any((i == 0 or s[i - 1] == slash) and _glob(p, s[i:], crossing) for i in range(len(s) + 1))
        if p[0] == 0x2A:  # '*'
            p = p[1:]
            for i in range(len(s) + 1):
                if _glob(p, s[i:], crossing):
                    return True
                if i < len(s) and s[i] == slash and not crossing:
                    return False
            return False
        if p[0] == 0x3F:  # '?'
            if not s or (s[0] == slash and not crossing):
                return False
            p, s = p[1:], s[1:]
            continue
        if not s or p[0] != s[0]:
            return False
        p, s = p[1:], s[1:]
    return s == b""


def ref_match(pattern: str, branch: str, data_center: bool) -> tuple[bool, bool]:
    """Match a branch restriction pattern against a branch name: (matched,
    supported). A refs/heads/ prefix is stripped, as Bitbucket does. Data
    Center patterns are Ant-style. Bitbucket Cloud does not document whether
    "*" crosses "/"; when the two readings disagree for this branch the match
    is reported unsupported, so hallpass never guesses. Character classes and
    alternations are unsupported on both."""
    pattern = pattern.removeprefix("refs/heads/")
    if any(c in pattern for c in "[]{}"):
        return False, False
    segment = glob(pattern, branch, False)
    # UNVERIFIED: whether a pattern without "/" also matches a branch of
    # that name inside a folder (main against release/main). When the two
    # readings differ, the match is unsupported.
    if "/" not in pattern and "/" in branch:
        any_dir = glob("**/" + pattern, branch, False)
        if any_dir != segment:
            return False, False
    if data_center:
        return segment, True
    # UNVERIFIED: whether Cloud's "*" matches across "/".
    crossing = glob(pattern, branch, True)
    if crossing != segment:
        return False, False
    return segment, True


def describe_level(have: Level, needed: Level) -> str:
    if have >= needed:
        return f"has {have} (needs {needed})"
    return f"has {have}, needs {needed}"
